src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,646 @@
+     * @param provider
+     *     the provider to use for this key derivation
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if the specified provider does not support the specified KDF
+     *     algorithm
+     * @throws NullPointerException
+     *     if the {@code algorithm} or {@code provider} is {@code null}
+     */
+    public static KDF getInstance(String algorithm, Provider provider)
+        throws NoSuchAlgorithmException {
+        Objects.requireNonNull(algorithm, "algorithm should not be null");
+        Objects.requireNonNull(provider, "provider should not be null");
+        try {
+            return getInstance(algorithm, null, provider);
+        } catch (InvalidAlgorithmParameterException e) {
+            throw new NoSuchAlgorithmException(
+                "No implementation found using null KDFParameters", e);
+        }
+    }
+
+    /**
+     * Returns a {@code KDF} instance initialized with the specified algorithm and
+     * is initialized with the specified parameters.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use.
+     *     See the {@code KDF} section in the <a href=
+     *     "{@docRoot}/../specs/security/standard-names.html#kdf-algorithms">
+     *     Java Security Standard Algorithm Names Specification</a>
+     *     for information about standard KDF algorithm names.
+     * @param kdfParameters
+     *     the {@code KDFParameters} used to configure the derivation
+     *     algorithm or {@code null} if no additional parameters are provided
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if no {@code Provider} supports a {@code KDFSpi} implementation for
+     *     the specified algorithm
+     * @throws InvalidAlgorithmParameterException
+     *     if the initialization parameters are inappropriate for this
+     *     {@code KDF} or if no provider can be found which supports the
+     *     parameters
+     * @throws NullPointerException
+     *     if the {@code algorithm} is {@code null}
+     */
+    public static KDF getInstance(String algorithm,
+                                  KDFParameters kdfParameters)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Objects.requireNonNull(algorithm, "algorithm should not be null");
+        // make sure there is at least one service from a signed provider
+        Iterator<Service> t = GetInstance.getServices("KDF", algorithm);
+        while (t.hasNext()) {
+            Service s = t.next();
+            if (!JceSecurity.canUseProvider(s.getProvider())) {
+                continue;
+            }
+            try {
+                Object obj = s.newInstance(kdfParameters);
+                if (!(obj instanceof KDFSpi spiObj)) {
+                    continue;
+                }
+                return new KDF(spiObj, s.getProvider(), t, algorithm, kdfParameters);
```

Comment:
If there is no other services supporting the requested KDF algorithm, we can construct the KDF object without `t` and lock in the found `spiObj` and its provider? Say
```
            if (t.hasNext()) {
                return new KDF(d, t, algorithm, kdfParameters);
            } else { // no other choices, lock down provider
                return new KDF(d, algorithm, kdfParameters);
            }
```

Reply:
Neat! I think in many cases there is only one provider. Maybe we can make the same enhancement to other classes.